### ISSUE TITLE
C2Rust Emits Unsafe Reference to Packed Field Causing Rust Error E0793

### ISSUE THREAD

**Issue by Yeaseen** [immunant/c2rust#1239]
C2Rust incorrectly emits a reference to a field inside a `#pragma pack(1) struct`, which results in an unsafe and invalid reference in Rust. This triggers Rust compiler error `E0793`, as references to potentially unaligned fields are undefined behavior in Rust.

## C code:

```c
#include <stdlib.h>
#pragma pack(push, 1)
struct S {
    int x;
};
#pragma pack(pop)

int func() {
    struct S* s = malloc(sizeof(struct S));
    return &s->x == &s->x;
}

int main(){
    return 0;
}
```

## Translated Rust Code:

```rust
#[derive(Copy, Clone)]
#[repr(C, packed)]
pub struct S {
    pub x: libc::c_int,
}
#[no_mangle]
pub unsafe extern "C" fn func() -> libc::c_int {
    let mut s: *mut S = malloc(::core::mem::size_of::<S>() as libc::c_ulong) as *mut S;
    return (&mut (*s).x as *mut libc::c_int == &mut (*s).x as *mut libc::c_int)
        as libc::c_int;
}
```

## Rust build error:

```text
error[E0793]: reference to packed field is unaligned
  --> src/runner.rs:22:13
   |
22 |     return (&mut (*s).x as *mut libc::c_int == &mut (*s).x as *mut libc::c_int)
   |             ^^^^^^^^^^^
   |
   = note: packed structs are only aligned by one byte, and many modern architectures penalize unaligned field accesses
   = note: creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
   = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)

error[E0793]: reference to packed field is unaligned
  --> src/runner.rs:22:48
   |
22 |     return (&mut (*s).x as *mut libc::c_int == &mut (*s).x as *mut libc::c_int)
   |                                                ^^^^^^^^^^^
   |
   = note: packed structs are only aligned by one byte, and many modern architectures penalize unaligned field accesses
   = note: creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
   = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)

For more information about this error, try `rustc --explain E0793`.
```

**Comment by jameysharp** [comment]
I just learned the other day that since Rust 1.51, the [addr_of!](https://doc.rust-lang.org/std/ptr/macro.addr_of.html) and [addr_of_mut!](https://doc.rust-lang.org/std/ptr/macro.addr_of_mut.html) macros should be used instead of this pattern of taking a reference and casting it to a raw pointer. As I understand it, that should fix this bug.

As of Rust 1.82, the macros are "soft-deprecated" in favor of the new `&raw const` and `&raw mut` syntax, which does the same thing. But since c2rust is currently targeting older versions of the compiler, I think the macros are the right choice for now.

**Comment by workingjubilee** [comment]
more than "does the same thing", the macros actually expanded to the syntax but it was unstable at the time (as std can use macros to "smuggle" unstable spans into Rust via use of compiler features). yeah, for expanded or generated code like c2rust where the prettiness isn't the _first_ concern, the macros should be fine.
